rfc/rfc-95/rfc-95.md

```diff
@@ -0,0 +1,319 @@
+  }
+  
+  @Override
+  public RecordsWithSplitIds<RecordAndPosition<T>> fetch() throws IOException {
+    metrics.incrementSplitReaderFetchCalls(1);
+    if (currentReader == null) {
+      HudiSourceSplit nextSplit = splits.poll();
+      if (nextSplit != null) {
+        currentSplit = nextSplit;
+        currentSplitId = nextSplit.splitId();
+        currentReader = readerContext.apply(currentSplit);
+      } else {
+        return new RecordsBySplits<>(Collections.emptyMap(), Collections.emptySet());
+      }
+    }
+
+    if (currentReader.hasNext()) {
+      // Because Iterator#next() doesn't support checked exception,
+      // we need to wrap and unwrap the checked IOException with UncheckedIOException
+      try {
+        return currentReader.next();
+      } catch (UncheckedIOException e) {
+        throw e.getCause();
+      }
+    } else {
+      return finishSplit();
+    }
+  }
+
+  @Override
+  public void handleSplitsChanges(SplitsChange<HudiSourceSplit> splitsChange) {
+    if (!(splitsChange instanceof SplitsAddition)) {
+      throw new UnsupportedOperationException(
+          String.format("Unsupported split change: %s", splitsChange.getClass()));
+    }
+
+    if (splitComparator != null) {
+      List<HudiSourceSplit> newSplits = Lists.newArrayList(splitsChange.splits());
+      newSplits.sort(splitComparator);
+      splits.addAll(newSplits);
+    } else {
+      splits.addAll(splitsChange.splits());
+    }
+  }
+
+  @Override
+  public void wakeUp() {
+  }
+
+  @Override
+  public void close() throws Exception {
+    currentSplitId = null;
+    if (currentReader != null) {
+      currentReader.close();
+    }
+  }
+
+  @Override
+  public void pauseOrResumeSplits(
+      Collection<HudiSourceSplit> splitsToPause, Collection<HudiSourceSplit> splitsToResume) {
+  }
+}
+```
+
+### Hudi Source Reader
+```java 
+public class HudiSourceReader<T>
+    extends SingleThreadMultiplexSourceReaderBase<
+        RecordAndPosition<T>, T, HudiSourceSplit, HudiSourceSplit> {
+  
+  // ... 
+  @Override
+  public void start() {
+    // We request a split only if we did not get splits during the checkpoint restore.
+    // Otherwise, reader restarts will keep requesting more and more splits.
+    if (getNumberOfCurrentlyAssignedSplits() == 0) {
+      requestSplit(Collections.emptyList());
+    }
+  }
+
+  @Override
+  protected void onSplitFinished(Map<String, HudiSourceSplit> finishedSplitIds) {
+    requestSplit(Lists.newArrayList(finishedSplitIds.keySet()));
+  }
+
+  @Override
+  protected HudiSourceSplit initializedState(HudiSourceSplit split) {
+    return split;
+  }
+
+  @Override
+  protected HudiSourceSplit toSplitType(String splitId, HudiSourceSplit splitState) {
+    return splitState;
+  }
+
+  private void requestSplit(Collection<String> finishedSplitIds) {
+    context.sendSourceEventToCoordinator(new SplitRequestEvent(finishedSplitIds));
+  }
+  
+  // ...
+}
+```
+
+### Hudi SplitEnumerator
```

Comment:
It'd be better to mention the checkpoint of source in this part.

Reply:
Added.